### PR TITLE
raftstore: mark peer as pending remove after destroying it

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

### What problem does this PR solve?

Problem Summary:
If a peer was destroyed, the subsequent proposals should be neglected.

### What is changed and how it works?

What's Changed:
Mark peer as `pending_remove` after destroying it.
### Related changes

- Need to cherry-pick to the release branch

### Check List 

- Unit test
- Integration test

### Release note 
* Reject handling any request on a destroyed peer.

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -106,8 +106,10 @@ pub struct PeerFsm<E: KvEngine> {
 
 impl<E: KvEngine> Drop for PeerFsm<E> {
     fn drop(&mut self) {
+        info!("peer fsm destroyed"; "region_id" => self.region_id());
         self.peer.stop();
         while let Ok(msg) = self.receiver.try_recv() {
+            info!("drop message after peer was destroyed"; "region_id" => self.region_id(), "msg" => ?msg);
             let callback = match msg {
                 PeerMsg::RaftCommand(cmd) => cmd.callback,
                 PeerMsg::CasualMessage(CasualMessage::SplitRegion { callback, .. }) => callback,
```
Comment (overvenus) [inline components/raftstore/src/store/fsm/peer.rs:115]: What about `SignificantMsg`s?

File: components/raftstore/src/store/fsm/peer.rs [hunk 2]
```diff
@@ -106,8 +106,10 @@ pub struct PeerFsm<E: KvEngine> {
 
 impl<E: KvEngine> Drop for PeerFsm<E> {
     fn drop(&mut self) {
+        info!("peer fsm destroyed"; "region_id" => self.region_id());
         self.peer.stop();
         while let Ok(msg) = self.receiver.try_recv() {
+            info!("drop message after peer was destroyed"; "region_id" => self.region_id(), "msg" => ?msg);
```
Comment (overvenus) [inline components/raftstore/src/store/fsm/peer.rs:112]: Please do not exceed 100 chars.

Comment (BusyJay) [reply components/raftstore/src/store/fsm/peer.rs:112]: I don't think the message is necessary. And `msg` may contains sensitive data.

File: components/raftstore/src/store/fsm/peer.rs [hunk 3]
```diff
@@ -106,8 +106,10 @@ pub struct PeerFsm<E: KvEngine> {
 
 impl<E: KvEngine> Drop for PeerFsm<E> {
     fn drop(&mut self) {
+        info!("peer fsm destroyed"; "region_id" => self.region_id());
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:109]: Too verbose, especially when a node is shutting down.

File: components/raftstore/src/store/fsm/peer.rs (hunk 2 above)
Comment (5kbpers) [reply components/raftstore/src/store/fsm/peer.rs:112]: 😅I will revert them.

**Comment by BusyJay** [comment]
When a peer is destroyed, all proposal will be cleared in drop method, isn't that sufficient enough? Besides, in this case `pending_remove` should be true.

**Comment by 5kbpers** [comment]
> When a peer is destroyed, all proposal will be cleared in drop method, isn't that sufficient enough? Besides, in this case `pending_remove` should be true.

As the test case has shown, the read request will succeed after the peer was destroyed, this will affect CDC since we cannot observe any event through.

**Comment by BusyJay** [comment]
Seems we should set `pending_remove` to true when destroying the peer. `pending_remove` and `stopped` are different. A `stopped` peer can be restart later. /cc @gengliqi

**Pull request merged**